core/src/util/fastmap.h

```diff
@@ -0,0 +1,115 @@
+#pragma once
+#include <vector>
+#include <string>
+#include <algorithm>
+
+namespace Tangram {
+
+// see also:
+// http://www.boost.org/doc/libs/1_59_0/doc/html/boost/container/flat_map.html
+// https://realm.io/assets/news/binary-search/blog.cpp
+
+template<typename K, typename T>
+struct fastmap {
+    std::vector<std::pair<K, T>> map;
+
+    using iterator = typename std::vector<std::pair<K, T>>::iterator;
+    using const_iterator = typename std::vector<std::pair<K, T>>::const_iterator;
+
+    T& operator[](const K& key) {
+        iterator it = std::lower_bound(
+            map.begin(), map.end(), key,
+            [&](const auto& item, const auto& key) {
+                return item.first < key;
+            });
+
+        if (it == map.end() || it->first != key) {
+            auto entry = map.emplace(it, key, T{});
+            return entry->second;
+        }
+
+        return it->second;
+    }
+
+    const_iterator find(const K& key) const {
+        const_iterator it = std::lower_bound(
+            map.begin(), map.end(), key,
+            [&](const auto& item, const auto& key) {
+                return item.first < key;
+            });
+
+        if (it == map.end() || it->first.k == key) {
```

Comment:
Ohh I get it. Dumb thing. :+1:
On Oct 29, 2015 12:10 AM, "Hannes Janetzek" <notifications@github.com>
wrote:

> In core/src/util/fastmap.h
> <https://github.com/tangrams/tangram-es/pull/357#discussion_r43347441>:
>
> > +        if (it == map.end() || it->first != key) {
> > +            auto entry = map.emplace(it, key, T{});
> > +            return entry->second;
> > +        }
> > +
> > +        return it->second;
> > +    }
> > +
> > +    const_iterator find(const K& key) const {
> > +        const_iterator it = std::lower_bound(
> > +            map.begin(), map.end(), key,
> > +            [&](const auto& item, const auto& key) {
> > +                return item.first < key;
> > +            });
> > +
> > +        if (it == map.end() || it->first.k == key) {
>
> but where does it->first look into when it == end? ;)
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/tangrams/tangram-es/pull/357/files#r43347441>.
>